### PR TITLE
JDK-8290020 Deadlock in leakprofiler::emit_events during shutdown

### PR DESCRIPTION
**Pull request by LudwikJaniuk**
Add a boolean parameter to Jfr::on_vm_shutdown to differentiate the "called from java" case, and in that case to not call JfrEmergencyDump::on_vm_shutdown.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8290020](https://bugs.openjdk.org/browse/JDK-8290020): Deadlock in leakprofiler::emit_events during shutdown

### Reviewers
 * [Markus Grönlund](https://openjdk.org/census#mgronlun) (@mgronlun - **Reviewer**)
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)
 * [Erik Gahlin](https://openjdk.org/census#egahlin) (@egahlin - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/9465/head:pull/9465` \
`$ git checkout pull/9465`

Update a local copy of the PR: \
`$ git checkout pull/9465` \
`$ git pull https://git.openjdk.org/jdk pull/9465/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9465`

View PR using the GUI difftool: \
`$ git pr show -t 9465`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/9465.diff">https://git.openjdk.org/jdk/pull/9465.diff</a>

</details>

### CODE REVIEW

**Comment by LudwikJaniuk** [comment]
trying to trigger jcheck...

**Review by mgronlun — CHANGES_REQUESTED** [review]
Good work Ludvig, thanks for fixing this.

File: test/jdk/jdk/jfr/jvm/TestDumpOnCrash.java [hunk 1]
```diff
@@ -76,41 +76,36 @@ public static void main(String[] args) throws Exception {
...
-            test(CrasherIllegalAccess.class, "", false, dumppath.toString());
-            test(CrasherHalt.class, "", true, dumppath.toString());
-            test(CrasherHalt.class, "", false, dumppath.toString());
+            test(CrasherIllegalAccess.class, "", true, dumppath.toString(), true);
+            test(CrasherIllegalAccess.class, "", false, dumppath.toString(), true);
         } finally {
             dumppath.toFile().delete();
         }
 
+        test(CrasherHalt.class, "", true, dumppath.toString(), false);
```
[the remark this answers is not in the document]
Comment (LudwikJaniuk) [reply test/jdk/jdk/jfr/jvm/TestDumpOnCrash.java]: Fine by me

**Comment by dholmes-ora** [comment]
What exactly is the constraint on calling `JfrEmergencyDump::on_vm_shutdown`? Is it that there must not be any Java frames on the stack? While it would be very unlikely in theory there is nothing to stop a daemon Java thread from calling a native method which calls JNI DestroyJavaVM, and so we get to `before_exit` in a "non-halt" case but the thread is no different to one that would call Runtime.halt().

**Comment by dholmes-ora** [comment]
> What exactly is the constraint on calling `JfrEmergencyDump::on_vm_shutdown`? Is it that there must not be any Java frames on the stack? While it would be very unlikely in theory there is nothing to stop a daemon Java thread from calling a native method which calls JNI DestroyJavaVM, and so we get to `before_exit` in a "non-halt" case but the thread is no different to one that would call Runtime.halt().

But if that happens we have worse problems anyway as there is no concurrency control for shutting down the VM that way.

**Review by dholmes-ora — APPROVED** [review]
It would be nicer if `JfrEmergencyDump::on_vm_shutdown` could detect whatever condition actually causes it a problem in the `halt()` case and do nothing; but I don't know how hard that is to detect.

The proposed workaround is okay.

Thanks.

**Review by egahlin — APPROVED** [review]

**Comment by dholmes-ora** [comment]
@mgronlun  or @egahlin do you have any feedback on my query above?

**Comment by egahlin** [comment]
> @mgronlun or @egahlin do you have any feedback on my query above?

The emergency dump was added because we could not execute Java code during JVM crash and the code to extract a recording file from a core dump was hard to maintain. 

The functionality was never meant to be triggered from Java.

One could argue the emergency dump should happen if somebody calls Runtime::halt(int), but I don't think that's what users want. Most likely they want a fast exit. If they call System.exit(int), JFR will do an ordinary dump in the shutdown hook, assuming JFR has been setup to dump on exit.

**Comment by mgronlun** [comment]
Runtime.halt() is tricky in that this method does not cause shutdown hooks to be started. If the shutdown sequence has already been initiated then this method does not wait for any running shutdown hooks to finish their work. This means it could deadlock with the normal Leakprofiler dump logic, issued either by the thread that stops a recording or if the shutdown hooks have already started, with the shutdown hook. It was an oversight to not exclude Runtime.halt() from calling into the emergency dump logic. One could argue that it could be distinguished in that it does not pass the exception handler, so we are not crashing. Unfortunately, that is only one part of the problem because we also attempt to emit the emergency dump in cases of out-of-memory, which is very hard to coordinate correctly.

**Pull request closed without merging**